integrations/opentelemetry-wsgi/src/opentelemetry_wsgi/__init__.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The opentelemetry_wsgi integration provides a WSGI middleware that can be used
+on any WSGI framework (such as Django / Flask) to track requests timing through
+OpenTelemetry.
+"""
+
+import opentelemetry.trace as trace
+
+
+class OpenTelemetryMiddleware:
+    """The WSGI application middleware.
+
+    This class is used to create and annotate spans for requests to a WSGI
+    application.
+
+    :param wsgi: The WSGI application callable.
+    """
+
+    def __init__(self, wsgi):
+        self.wsgi = wsgi
+        self.start_response = None
+        self.span = None
+
+    def _add_request_attributes(self, environ):
+        self.span.add_attribute("http.method", environ["HTTP_METHOD"])
+        self.span.add_attribute("http.path", environ["PATH_INFO"])
+        self.span.add_attribute("http.route", environ["PATH_INFO"])
+
+        host = environ.get("HTTP_HOST") or environ.get("SERVER_NAME")
+        if host is not None:
+            self.span.add_attribute("http.host", host)
+
+        url = environ.get("REQUEST_URI") or environ.get("RAW_URI")
+        if url is not None:
+            self.span.add_attribute("http.url", url)
+
+    def _add_response_attributes(self, status, response_headers, *args):
+        try:
+            status_code = int(status.split(" ", 1)[0])
+            self.span.add_attribute("http.status_code", status_code)
+        except ValueError:
+            pass
+
+        return self.start_response(status, response_headers, *args)
+
+    def __call__(self, environ, start_response):
+        """The WSGI application
+
+        :param environ: A WSGI environment.
+        :param start_response: The WSGI start_response callable.
+        """
+        self.start_response = start_response
+
+        tracer = trace.tracer()
+        method = environ["REQUEST_METHOD"]
+        path = environ.get("PATH_INFO", "<unknown>")
+
+        span_name = "[{}]{}".format(method, path)
+
+        with tracer.start_span(span_name) as span:
+            self.span = span
+            # TODO: enable request attributes after set_attribute API is
+            # implemented
+            # self._add_request_attributes(environ)
+
+            # TODO: change start_response to self._add_response_attributes after
+            # set_attribute API is implemented
+            for yielded in self.wsgi(environ, start_response):
```

Comment:
This is definitely an interesting problem that I haven't considered in depth yet. From reading the specification I believe the optimization you're referring to is this part of the spec?

> The callable must return an iterable object, and must not perform any data transmission until and unless the server/gateway actually receives the iterable as a return value from the application. (To do otherwise would prevent middleware from being able to interpret or override the response data.)

From a functional perspective, I believe the behavior is correct since the loop is iterating over the `wsgi.file_wrapper` iterable.

From a timing perspective, I wonder what the response time should be recorded as in this case? Should it include the time the server spends streaming the file? For any other request, the answer would generally be yes.

Performance-wise, the intent was to avoid allocating a new response-like iterable object and instead pass the values through transparently, taking advantage of the context manager behavior for proper timing. In order to maintain the pass-through behavior, the application would have to be invoked and a response wrapper object that was iterable would have to be created.

```python
def __call__(self, environ, start_response):
    iterable = self.wsgi(environ, start_response)
    return ResponseTimer(iterable)  # allocation of the ResponseTimer object
```

Reply:
> Maybe more relevant, some servers could optimize if an object with known len is returned.

This is a very good point. Some WSGI applications return a string directly which results in iterating over the entire length of the string.  The middleware should definitely not remove optimizations around applications that return a string type rather than a contained iterable like a list or tuple.

Reply:
Also, as far as I can tell, the only way to preserve the `wsgi.file_wrapper` optimization is to [add an isinstance check](https://github.com/benoitc/gunicorn/blob/b12d36409546c20347c9501e604ff42368a73af9/gunicorn/workers/sync.py#L177) for the `wsgi.file_wrapper`. Using this approach implies that we cannot time the full response time where the file contents are streamed.

So it sounds like the trade-off is performance versus accuracy of response time values.